sda/router_test.go

```diff
@@ -0,0 +1,325 @@
+func TestTcpRouterNew(t *testing.T) {
+	h1 := NewMockTcpRouter(2000)
+	if h1 == nil {
+		t.Fatal("Couldn't setup a Host")
+	}
+	err := h1.Close()
+	if err != nil {
+		t.Fatal("Couldn't close", err)
+	}
+}
+
+// Test closing and opening of Host on same address
+func TestTcpRouterClose(t *testing.T) {
+	h1 := NewMockTcpRouter(2000)
+	h2 := NewMockTcpRouter(2001)
+	go h1.Run()
+	_, err := h2.Connect(h1.serverIdentity)
+	if err != nil {
+		t.Fatal("Couldn't Connect()", err)
+	}
+	err = h1.Close()
+	if err != nil {
+		t.Fatal("Couldn't close:", err)
+	}
+	err = h2.Close()
+	if err != nil {
+		t.Fatal("Couldn't close:", err)
+	}
+	log.Lvl3("Finished first connection, starting 2nd")
+	h3 := NewTCPHost(2002)
+	go h3.Run()
+	c, err := h2.Connect(h3.ServerIdentity)
+	if err != nil {
+		t.Fatal(h2, "Couldn Connect() to", h3)
+	}
+	log.Lvl3("Closing h3")
+	err = h3.Close()
+	if err != nil {
+		// try closing the underlying connection manually and fail
+		c.Close()
+		t.Fatal("Couldn't Close()", h3)
+	}
+}
+
+// Test if TCPRouter fits the interface such as calling Run(), then Close(),
+// should return
+func TestTcpRouterRunClose(t *testing.T) {
+	h := NewMockTcpRouter(2000)
+	var stop = make(chan bool)
+	go func() {
+		stop <- true
+		h.Run()
+		stop <- true
+	}()
+	<-stop
+	// Time needed so the listener is up. Equivalent to "connecting ourself" as
+	// we had before.
+	time.Sleep(500 * time.Millisecond)
+	h.Close()
+	select {
+	case <-stop:
+		return
+	case <-time.After(500 * time.Millisecond):
+		t.Fatal("TcpHost should have returned from Run() by now")
+	}
+}
+
+func TestTcpRouterReconnection(t *testing.T) {
+	h1 := NewMockTcpRouter(2000)
+	h2 := NewMockTcpRouter(2001)
+	/* h1 := NewLocalHost(2000)*/
+	/*h2 := NewLocalHost(2001)*/
+	defer h1.Close()
+	defer h2.Close()
+
+	go h1.Run()
+	go h2.Run()
+
+	log.Lvl1("Sending h1->h2")
+	log.ErrFatal(sendrcv_proc(h1, h2))
+	log.Lvl1("Sending h2->h1")
+	log.ErrFatal(sendrcv_proc(h2, h1))
+	log.Lvl1("Closing h1")
+	assert.Nil(t, h1.closeConnections())
+
+	log.Lvl1("Listening again on h1")
+	go h1.listen()
+	time.Sleep(200 * time.Millisecond)
+	log.Lvl1("Sending h2->h1")
+	log.ErrFatal(sendrcv_proc(h2, h1))
+	log.Lvl1("Sending h1->h2")
+	log.ErrFatal(sendrcv_proc(h1, h2))
+
+	log.Lvl1("Shutting down listener of h2")
+
+	// closing h2, but simulate *hard* failure, without sending a FIN packet
+	// XXX Actually it DOES send a FIN packet: using tcphost.Close(), it closes
+	// the listener AND all the connections (calling golang tcp connection
+	// Close() which I'm pretty sure will send a FIN packet)
+	// This test is ambiguous as it does not really simulate a network hardware
+	// failure of a node, but merely a host which does weird abort
+	// connections...
+	// One idea if we really want to simulate that is calling tcphost.Close()
+	// and at the same time, at the IP level, blocking all FIN packet.
+	// Then start a new host with the same entity etc..
+	// See also https://github.com/tylertreat/comcast
```

Comment:
Or at least rename it. Because now it doesn't attempt to reconnect (`TestTcpRouterReconnection `), right?